### ISSUE TITLE
Infinite recursion error if a field has the same name as its type

### ISSUE THREAD

**Issue by guillaume-alliander** [pydantic/pydantic#8240]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

If a field has the same name as its type, and is defined with `Field()`, then a `RecursionError` happens, 

In the example below:
- this is true with pydantic `dataclass` or `BaseModel`
- using the default python dataclass, this does not happen
- if `=Field()` is not given, this does not happen
- it seems to only happen if the type of the field is another Pydantic object (`int: int` or `Path:Path` with `from pathlib import Path` are both fine)

The given snippet is enough to blow python up. 

My expectation would be to see "42" if I add:

```python
o=Other(Thing=Thing(i=42))
print(o.Thing.i)
```

Which is the case if I use python's dataclasses. 

### Example Code

```Python
# from dataclasses import dataclass # works
from pydantic import Field
from pydantic.dataclasses import dataclass  # blows up


@dataclass
class Thing:
    i:int

@dataclass
class Other:
    Thing: Thing=Field() # Works if Field is no present
```

### Python, Pydantic & OS Version

```Text
python -c "import pydantic.version; print(pydantic.version.version_info())"                                                             0s - 11:48:19
             pydantic version: 2.5.0
        pydantic-core version: 2.14.1
          pydantic-core build: profile=release pgo=true
                 install path: /home/gui/.conda/envs/pycgmes10/lib/python3.10/site-packages/pydantic
               python version: 3.10.11 (main, Apr 20 2023, 19:02:41) [GCC 11.2.0]
                     platform: Linux-5.10.102.1-microsoft-standard-WSL2-x86_64-with-glibc2.35
             related packages: pyright-1.1.335 typing_extensions-4.8.0 mypy-1.5.1
```

**Comment by Viicos** [comment]
Duplicate of: https://github.com/pydantic/pydantic/issues/7327 (you will find explanation and related issues here).
I'll take a look to see if it's possible to have a better error message here as this issue comes up pretty often.

**Comment by sydney-runkle** [comment]
Marking as fixed by #8243
